File: Baekjoon/Class2/33.10845 큐.java
import java.io.*;
import java.util.NoSuchElementException;
import java.util.StringTokenizer;

class Main {
    private static class Queue<T> { // 연결 리스트 큐
        private static class Node<T> {  // 연결 리스트 노드
            T data;
            Node<T> next;
            public Node(T data, Node<T> next) {
                this.data = data;
                this.next = next;
            }
        }
        private Node<T> front; // 첫 노드
        private Node<T> back; // 마지막 노드
        private int size;

        public Queue() {
            front = back = null;
            size = 0;
        }
        public void push(T data) {
            size++;
            Node<T> nodeToPush = new Node<T>(data, null);
            if (size == 1)
                front = back = nodeToPush;
            else {
                back.next = nodeToPush;
                back = nodeToPush;
            }
        }
        public T pop() {
            if (empty()) throw new NoSuchElementException();
            size--;
            T data = front.data;
            front = front.next;
            if (front == null) back = null;
            return data;
        }
        public int size() {
            return size;
        }
        public boolean empty() {
            return size == 0;
        }
        public T front() {
            if (empty()) throw new NoSuchElementException();
            return front.data;
        }
        public T back() {
            if (empty()) throw new NoSuchElementException();
            return back.data;
        }
    }

    public static void main(String[] args) throws IOException {
        int n;
        Queue<Integer> nums = new Queue<>();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        for(int i = 0; i < n; i++) {
            String command;
            st = new StringTokenizer(br.readLine());
            command = st.nextToken();
            try {
                switch (command) {
                    case "push":
                        nums.push(Integer.parseInt(st.nextToken()));
                        break;
                    case "pop":
                        sb.append(nums.pop()).append('\n');
                        break;
                    case "size":
                        sb.append(nums.size()).append('\n');
                        break;
                    case "empty":
                        sb.append(nums.empty() ? 1 : 0).append('\n');
                        break;
                    case "front":
                        sb.append(nums.front()).append('\n');
                        break;
                    case "back":
                        sb.append(nums.back()).append('\n');
                        break;
                }
            }
            catch (NoSuchElementException e) { // 큐 빔
                sb.append("-1\n");
            }
        }
        System.out.print(sb);
        br.close();
    }
}

/* 라이브러리 사용
import java.io.*;
import java.util.*;

class Main {
    public static void main(String[] args) throws IOException {
        int n;
        Deque<Integer> nums = new ArrayDeque<>();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        for(int i = 0; i < n; i++) {
            String command;
            st = new StringTokenizer(br.readLine());
            command = st.nextToken();
            switch (command) {
                case "push":
                    nums.addLast(Integer.parseInt(st.nextToken()));
                    break;
                case "pop":
                    if(nums.isEmpty()) {
                        sb.append("-1\n");
                        continue;
                    }
                    sb.append(nums.removeFirst()).append('\n');
                    break;
                case "size":
                    sb.append(nums.size()).append('\n');
                    break;
                case "empty":
                    sb.append(nums.isEmpty() ? 1 : 0).append('\n');
                    break;
                case "front":
                    if(nums.isEmpty()) {
                        sb.append("-1\n");
                        continue;
                    }
                    sb.append(nums.peekFirst()).append('\n');
                    break;
                case "back":
                    if(nums.isEmpty()) {
                        sb.append("-1\n");
                        continue;
                    }
                    sb.append(nums.peekLast()).append('\n');
                    break;
            }
        }
        System.out.print(sb);
        br.close();
    }
}
*/

File: Baekjoon/Class2/33.10845 큐.cpp
#include <iostream>
#include <string>
using namespace std;

template <typename T> // 연결 리스트 큐
class Queue {
private:
    struct Node { // 연결 리스트 노드
        T data;
        Node* next;
    };
    Node* _front; // 첫 노드
    Node* _back; // 마지막 노드
    int _size;
public:
    Queue() : _front(nullptr), _back(nullptr), _size(0) {}
    ~Queue() {
        while (!empty())
            pop();
    }
    void push(const T& data) {
        _size++;
        Node* node_to_push = new Node;
        node_to_push->data = data;
        node_to_push->next = nullptr;
        if (_size == 1)
            _front = _back = node_to_push;
        else {
            _back->next = node_to_push;
            _back = node_to_push;
        }
    }
    T pop() {
        if (empty()) throw out_of_range("queue is empty");
        _size--;
        Node* node_to_pop = _front;
        T data = node_to_pop->data;
        _front = node_to_pop->next;
        delete node_to_pop;
        if (!_front) _back = nullptr;
        return data;
    }
    int size() {
        return _size;
    }
    bool empty() {
        return _size == 0;
    }
    T front() {
        if (empty()) throw out_of_range("queue is empty");
        return _front->data;
    }
    T back() {
        if (empty()) throw out_of_range("queue is empty");
        return _back->data;
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n;
    Queue<int> nums;

    cin >> n;
    while (n--) {
        string command;
        cin >> command;
        try {
            if (command == "push") {
                int num;
                cin >> num;
                nums.push(num);
            }
            else if (command == "pop")
                cout << nums.pop() << '\n';
            else if (command == "size")
                cout << nums.size() << '\n';
            else if (command == "empty")
                cout << nums.empty() << '\n';
            else if (command == "front")
                cout << nums.front() << '\n';
            else if (command == "back")
                cout << nums.back() << '\n';
        }
        catch (const out_of_range& e) { // 큐 빔
            cout << -1 << '\n';
        }
    }
    return 0;
}

/* 라이브러리 사용
#include <iostream>
#include <string>
#include <queue>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n;
    queue<int> nums;

    cin >> n;
    while (n--) {
        string command;
        cin >> command;
        if (command == "push") {
            int num;
            cin >> num;
            nums.push(num);
        }
        else if (command == "pop") {
            if (nums.empty()) {
                cout << "-1\n";
                continue;
            }
            cout << nums.front() << '\n';
            nums.pop();
        }
        else if (command == "size")
            cout << nums.size() << '\n';
        else if (command == "empty")
            cout << nums.empty() << '\n';
        else if (command == "front") {
            if (nums.empty()) {
                cout << "-1\n";
                continue;
            }
            cout << nums.front() << '\n';
        }
        else if (command == "back") {
            if (nums.empty()) {
                cout << "-1\n";
                continue;
            }
            cout << nums.back() << '\n';
        }
    }
    return 0;
}
*/